### DISCUSSION TITLE
Help translating a PHP method to ZEPHIR

### DISCUSSION THREAD

**Discussion by henrywood** [zephir-lang/zephir#2415] [Q&A]
I have this PHP method:
```php
/**
	 * Converts array of UTF-8 characters to UTF16-BE string.<br>
	 * Based on: http://www.faqs.org/rfcs/rfc2781.html
	 * <pre>
	 *   Encoding UTF-16:
	 *
	 *   Encoding of a single character from an ISO 10646 character value to
	 *    UTF-16 proceeds as follows. Let U be the character number, no greater
	 *    than 0x10FFFF.
	 *
	 *    1) If U < 0x10000, encode U as a 16-bit unsigned integer and
	 *       terminate.
	 *
	 *    2) Let U' = U - 0x10000. Because U is less than or equal to 0x10FFFF,
	 *       U' must be less than or equal to 0xFFFFF. That is, U' can be
	 *       represented in 20 bits.
	 *
	 *    3) Initialize two 16-bit unsigned integers, W1 and W2, to 0xD800 and
	 *       0xDC00, respectively. These integers each have 10 bits free to
	 *       encode the character value, for a total of 20 bits.
	 *
	 *    4) Assign the 10 high-order bits of the 20-bit U' to the 10 low-order
	 *       bits of W1 and the 10 low-order bits of U' to the 10 low-order
	 *       bits of W2. Terminate.
	 *
	 *    Graphically, steps 2 through 4 look like:
	 *    U' = yyyyyyyyyyxxxxxxxxxx
	 *    W1 = 110110yyyyyyyyyy
	 *    W2 = 110111xxxxxxxxxx
	 * </pre>
	 * @param array $unicode array containing UTF-8 unicode values
	 * @param boolean $setbom if true set the Byte Order Mark (BOM = 0xFEFF)
	 * @return string
	 * @protected
	 * @author Nicola Asuni
	 * @since 2.1.000 (2008-01-08)
	 * @public static
	 */
	public static function arrUTF8ToUTF16BE($unicode, $setbom=false) {
		$outstr = ''; // string to be returned
		if ($setbom) {
			$outstr .= "\xFE\xFF"; // Byte Order Mark (BOM)
		}
		foreach ($unicode as $char) {
			if ($char == 0x200b) {
				// skip Unicode Character 'ZERO WIDTH SPACE' (DEC:8203, U+200B)
			} elseif ($char == 0xFFFD) {
				$outstr .= "\xFF\xFD"; // replacement character
			} elseif ($char < 0x10000) {
				$outstr .= chr($char >> 0x08);
				$outstr .= chr($char & 0xFF);
			} else {
				$char -= 0x10000;
				$w1 = 0xD800 | ($char >> 0x0a);
				$w2 = 0xDC00 | ($char & 0x3FF);
				$outstr .= chr($w1 >> 0x08);
				$outstr .= chr($w1 & 0xFF);
				$outstr .= chr($w2 >> 0x08);
				$outstr .= chr($w2 & 0xFF);
			}
		}
		return $outstr;
	}
´´´ 
	
	I am trying to convert it to ZEPHIR - This is my ZEPHIR method:
	
´´´zep
	        public static function arrUTF8ToUTF16BE(var unicode, var setbom = false)
        {
                var outstr, w1, w2, highSurrogate, lowSurrogate, counter, utfChar, temp;
                let outstr = ""; // string to be returned
                if (setbom) {
                        let outstr .= "\xFE\xFF"; // Byte Order Mark (BOM)
                }

                let counter = 0; // Initialize a counter

                while counter < count(unicode) { // Check if the counter is within bounds
                        let utfChar = unicode[counter]; // Get the character

                        if (utfChar == 0x200b) {
                                // skip Unicode Character "ZERO WIDTH SPACE" (DEC:8203, U+200B)
                        } elseif (utfChar == 0xFFFD) {
                                let outstr .= "\xFF\xFD"; // replacement character
                        } elseif (utfChar < 0x10000) {
                                let outstr .= chr(utfChar >> 0x08);
                                let outstr .= chr(utfChar & 0xFF);
                        } else {
                                let temp = utfChar - 0x10000;
                                let highSurrogate = (0xD800 | (temp >> 0x0a));
                                let lowSurrogate = (0xDC00 | (temp & 0x3FF));

                                let w1 = highSurrogate;
                                let w2 = lowSurrogate;

                                let outstr .= chr(w1 >> 0x08);
                                let outstr .= chr(w1 & 0xFF);
                                let outstr .= chr(w2 >> 0x08);
                                let outstr .= chr(w2 & 0xFF);
                        }

                        let counter++; // Increment the counter
                }

                return outstr;
        }
´´´
        
        However, I am told that these lines:
```zep   
    let temp = utfChar - 0x10000;
    let highSurrogate = (0xD800 | (temp >> 0x0a));
    let lowSurrogate = (0xDC00 | (temp & 0x3FF));
```    
    contain syntax errors ?
    
    How can I make my method work / compile ?
    
    Thanks

**Comment by mdhifta** [comment]
hello, I have read your code. In PHP there is no syntax with var, let, or others. they all start with $ when you want to define a variable. This is not JavaScript or languages ​​like Golang, Python and Java where when you want to create a variable you have to first declare the data type.

**Reply by henrywood** [reply]
It is *ZEPHIR* code ? I am trying to translate PHP to ZEPHIR so this code is ZEPHIR:

        public static function arrUTF8ToUTF16BE(var unicode, var setbom = false)
        {
                var outstr, w1, w2, highSurrogate, lowSurrogate, counter, utfChar, temp;
                let outstr = ""; // string to be returned
                if (setbom) {
                        let outstr .= "\xFE\xFF"; // Byte Order Mark (BOM)
                }

                let counter = 0; // Initialize a counter

                while counter < count(unicode) { // Check if the counter is within bounds
                        let utfChar = unicode[counter]; // Get the character

                        if (utfChar == 0x200b) {
                                // skip Unicode Character "ZERO WIDTH SPACE" (DEC:8203, U+200B)
                        } elseif (utfChar == 0xFFFD) {
                                let outstr .= "\xFF\xFD"; // replacement character
                        } elseif (utfChar < 0x10000) {
                                let outstr .= chr(utfChar >> 0x08);
                                let outstr .= chr(utfChar & 0xFF);
                        } else {
                                let temp = utfChar - 0x10000;
                                let highSurrogate = (0xD800 | (temp >> 0x0a));
                                let lowSurrogate = (0xDC00 | (temp & 0x3FF));

                                let w1 = highSurrogate;
                                let w2 = lowSurrogate;

                                let outstr .= chr(w1 >> 0x08);
                                let outstr .= chr(w1 & 0xFF);
                                let outstr .= chr(w2 >> 0x08);
                                let outstr .= chr(w2 & 0xFF);
                        }

                        let counter++; // Increment the counter
                }

                return outstr;
        }
        
        (but the problem is that it contains syntax errors on these lines:
            let temp = utfChar - 0x10000;
    let highSurrogate = (0xD800 | (temp >> 0x0a));
    let lowSurrogate = (0xDC00 | (temp & 0x3FF));

**Reply by yarcowang** [reply]
@henrywood 

I think it would be better to use [mb_convert_encoding](https://www.php.net/manual/en/function.mb-convert-encoding) than implement it again in zephir. Modern PHP all have `mbstring` extension embed.
```
        let unicode = mb_convert_encoding(unicode, "utf-8", "utf-16be");
                                                $to_encoding $from_encoding
```
It also works in calling zephir context.

**Reply by henrywood** [reply]
Thank you - will give it a try !

**Reply by yarcowang** [reply]
@henrywood 
Oh, take care of the parameters' order. I make it wrong. The 2nd is `$to_encoding`. You might need to switch them.
